File: integration-tests/java/delete-contained-entities/Session.java
// integration-tests/java/delete-contained-entities -- session. Direct Java
// port of c/delete-contained-entities/src/session.c -- see that file's
// header comment for the full scenario rationale and
// docs/design/integration-test-tier.md for the scenario spec.
//
// Required stdout markers: "Create topic:" x3, "Create writer for topic:"
// x2, "Create reader for topic:" x2, "Session: torn down via
// delete_contained_entities." Any failure path prints a line starting
// "FAIL:" and exits nonzero.

import io.zzdds.dcps.Dcps;
import io.zzdds.ext.Zzdds;

import java.util.Collections;
import java.util.concurrent.atomic.AtomicBoolean;

public class Session {
    static final int SAMPLE_COUNT = 5;
    static final int RECEIVE_TARGET = 2;
    static final int READER_READY_TIMEOUT_MS = 10000;
    static final int RECEIVE_TIMEOUT_MS = 20000;
    static final int WAIT_STEP_SEC = 1;
    static final int POLL_PERIOD_MS = 20;

    static final AtomicBoolean tornDown = new AtomicBoolean(false);

    static class WriterSyncState {
        final AtomicBoolean readerReady = new AtomicBoolean(false);
    }

    static int parseDomain(String[] args) {
        for (int i = 0; i < args.length - 1; i++) {
            if (args[i].equals("-d") || args[i].equals("--domain")) {
                return Integer.parseInt(args[i + 1]);
            }
        }
        return 0;
    }

    static Zzdds.zzdds.DataWriterListenerEx makeWriterListener(WriterSyncState state) {
        return new Zzdds.zzdds.DataWriterListenerEx() {
            public void on_offered_deadline_missed(Dcps.DDS.DataWriter w, Dcps.DDS.OfferedDeadlineMissedStatus s) {}
            public void on_offered_incompatible_qos(Dcps.DDS.DataWriter w, Dcps.DDS.OfferedIncompatibleQosStatus s) {}
            public void on_liveliness_lost(Dcps.DDS.DataWriter w, Dcps.DDS.LivelinessLostStatus s) {}

            public void on_publication_matched(Dcps.DDS.DataWriter w, Dcps.DDS.PublicationMatchedStatus s) {
                if (tornDown.get()) {
                    System.err.println("FAIL: listener fired after delete_contained_entities");
                    System.exit(1);
                }
            }

            public void on_reliable_reader_ready(int readerHandle, boolean isReady) {
                if (isReady) state.readerReady.set(true);
            }
        };
    }

    static Dcps.DDS.DataReaderListener makeReaderListener() {
        return new Dcps.DDS.DataReaderListener() {
            public void on_requested_deadline_missed(Dcps.DDS.DataReader r, Dcps.DDS.RequestedDeadlineMissedStatus s) {}
            public void on_requested_incompatible_qos(Dcps.DDS.DataReader r, Dcps.DDS.RequestedIncompatibleQosStatus s) {}
            public void on_sample_rejected(Dcps.DDS.DataReader r, Dcps.DDS.SampleRejectedStatus s) {}
            public void on_liveliness_changed(Dcps.DDS.DataReader r, Dcps.DDS.LivelinessChangedStatus s) {}
            public void on_data_available(Dcps.DDS.DataReader r) {}
            public void on_sample_lost(Dcps.DDS.DataReader r, Dcps.DDS.SampleLostStatus s) {}

            public void on_subscription_matched(Dcps.DDS.DataReader r, Dcps.DDS.SubscriptionMatchedStatus s) {
                if (tornDown.get()) {
                    System.err.println("FAIL: listener fired after delete_contained_entities");
                    System.exit(1);
                }
            }
        };
    }

    public static void main(String[] args) throws Exception {
        int domainId = parseDomain(args);

        Dcps.DDS.DomainParticipantFactory factory =
            (Dcps.DDS.DomainParticipantFactory) io.zzdds.runtime.ZzddsRuntime.createFactory();
        if (factory == null) {
            System.err.println("FAIL: createFactory() failed");
            System.exit(1);
        }

        Dcps.DDS.DomainParticipant dp = factory.create_participant(domainId, null, null, 0);
        if (dp == null) {
            System.err.println("FAIL: create_participant() failed on domain " + domainId);
            System.exit(1);
        }

        if (SessionEventTypeSupport.register(dp, "SessionEvent") != 0) {
            System.err.println("FAIL: register SessionEventTypeSupport failed");
            System.exit(1);
        }

        Dcps.DDS.Topic out1Topic = dp.create_topic("SessionOut1", "SessionEvent", null, null, 0);
        Dcps.DDS.Topic out2Topic = dp.create_topic("SessionOut2", "SessionEvent", null, null, 0);
        Dcps.DDS.Topic inTopic = dp.create_topic("SessionIn", "SessionEvent", null, null, 0);
        if (out1Topic == null || out2Topic == null || inTopic == null) {
            System.err.println("FAIL: create_topic() failed");
            System.exit(1);
        }
        System.out.println("Create topic: SessionOut1");
        System.out.println("Create topic: SessionOut2");
        System.out.println("Create topic: SessionIn");

        Dcps.DDS.Publisher pub = dp.create_publisher(null, null, 0);
        if (pub == null) {
            System.err.println("FAIL: create_publisher() failed");
            System.exit(1);
        }

        Dcps.DDS.DataWriterQos dwQos = new Dcps.DDS.DataWriterQos();
        pub.get_default_datawriter_qos(dwQos);
        dwQos.get_reliability().set_kind(Dcps.DDS.ReliabilityQosPolicyKind.RELIABLE_RELIABILITY_QOS);
        dwQos.get_history().set_kind(Dcps.DDS.HistoryQosPolicyKind.KEEP_ALL_HISTORY_QOS);

        Dcps.DDS.DataWriter out1Dw = pub.create_datawriter(out1Topic, dwQos, null, 0);
        Dcps.DDS.DataWriter out2Dw = pub.create_datawriter(out2Topic, dwQos, null, 0);
        if (out1Dw == null || out2Dw == null) {
            System.err.println("FAIL: create_datawriter() failed");
            System.exit(1);
        }
        System.out.println("Create writer for topic: SessionOut1");
        System.out.println("Create writer for topic: SessionOut2");

        WriterSyncState out1State = new WriterSyncState();
        WriterSyncState out2State = new WriterSyncState();
        Zzdds.zzdds.DataWriter zOut1Dw = (Zzdds.zzdds.DataWriter) io.zzdds.runtime.ZzddsRuntime.asZzddsDataWriter(out1Dw);
        Zzdds.zzdds.DataWriter zOut2Dw = (Zzdds.zzdds.DataWriter) io.zzdds.runtime.ZzddsRuntime.asZzddsDataWriter(out2Dw);
        if (zOut1Dw == null || zOut2Dw == null) {
            System.err.println("FAIL: asZzddsDataWriter() failed");
            System.exit(1);
        }
        if (zOut1Dw.set_listener_ex(makeWriterListener(out1State), Dcps.DDS.PUBLICATION_MATCHED_STATUS.value) != 0 ||
            zOut2Dw.set_listener_ex(makeWriterListener(out2State), Dcps.DDS.PUBLICATION_MATCHED_STATUS.value) != 0)
        {
            System.err.println("FAIL: set_listener_ex (writer) failed");
            System.exit(1);
        }

        Dcps.DDS.Subscriber sub = dp.create_subscriber(null, null, 0);
        if (sub == null) {
            System.err.println("FAIL: create_subscriber() failed");
            System.exit(1);
        }

        Dcps.DDS.DataReaderQos drQos = new Dcps.DDS.DataReaderQos();
        sub.get_default_datareader_qos(drQos);
        drQos.get_reliability().set_kind(Dcps.DDS.ReliabilityQosPolicyKind.RELIABLE_RELIABILITY_QOS);
        drQos.get_history().set_kind(Dcps.DDS.HistoryQosPolicyKind.KEEP_ALL_HISTORY_QOS);

        Dcps.DDS.DataReader inDr = sub.create_datareader(inTopic, drQos, null, 0);
        if (inDr == null) {
            System.err.println("FAIL: create_datareader(SessionIn) failed");
            System.exit(1);
        }
        System.out.println("Create reader for topic: SessionIn");

        // Exercise CFT-collateral cleanup under the cascade -- this project
        // has a real CFT bug history (see docs/decisions.md). Trivial
        // filter: just needs to exist and be attached, not to actually
        // narrow anything.
        Dcps.DDS.ContentFilteredTopic cft = dp.create_contentfilteredtopic("SessionIn_cft", inTopic, "seq >= 0", Collections.emptyList());
        if (cft == null) {
            System.err.println("FAIL: create_contentfilteredtopic() failed");
            System.exit(1);
        }
        Dcps.DDS.DataReader cftDr = sub.create_datareader(cft, drQos, null, 0);
        if (cftDr == null) {
            System.err.println("FAIL: create_datareader(SessionIn_cft) failed");
            System.exit(1);
        }
        System.out.println("Create reader for topic: SessionIn_cft");

        Dcps.DDS.DataReaderListener readerListener = makeReaderListener();
        if (inDr.set_listener(readerListener, Dcps.DDS.SUBSCRIPTION_MATCHED_STATUS.value) != 0 ||
            cftDr.set_listener(readerListener, Dcps.DDS.SUBSCRIPTION_MATCHED_STATUS.value) != 0)
        {
            System.err.println("FAIL: set_listener (reader) failed");
            System.exit(1);
        }

        Dcps.DDS.WaitSet ws = (Dcps.DDS.WaitSet) io.zzdds.runtime.ZzddsRuntime.createWaitSet();
        if (ws == null) {
            System.err.println("FAIL: createWaitSet() failed");
            System.exit(1);
        }
        Dcps.DDS.ReadCondition inRc = inDr.create_readcondition(
            Dcps.DDS.ANY_SAMPLE_STATE.value, Dcps.DDS.ANY_VIEW_STATE.value, Dcps.DDS.ANY_INSTANCE_STATE.value);
        if (inRc == null) {
            System.err.println("FAIL: create_readcondition() failed");
            System.exit(1);
        }
        if (ws.attach_condition(inRc) != 0) {
            System.err.println("FAIL: attach_condition() failed");
            System.exit(1);
        }

        SessionEventDataWriter out1Writer = new SessionEventDataWriter(out1Dw);
        SessionEventDataWriter out2Writer = new SessionEventDataWriter(out2Dw);
        SessionEventDataReader inReader = new SessionEventDataReader(inDr);

        // Gate writes on the peer's reader actually being registered, not
        // just matched -- see docs/design/integration-test-tier.md's
        // raw-loan/coherent-sets precedent and this project's
        // on_reliable_reader_ready work: matched-count alone (SEDP
        // discovery) does not imply the remote RELIABLE reader proxy has
        // registered this writer yet.
        long deadline = System.currentTimeMillis() + READER_READY_TIMEOUT_MS;
        while (!(out1State.readerReady.get() && out2State.readerReady.get())) {
            if (System.currentTimeMillis() > deadline) {
                System.err.println("FAIL: no reliable reader became ready within " + (READER_READY_TIMEOUT_MS / 1000) + "s");
                System.exit(1);
            }
            Thread.sleep(POLL_PERIOD_MS);
        }

        for (int seq = 0; seq < SAMPLE_COUNT; seq++) {
            Session_event.SessionEvent ev = new Session_event.SessionEvent();
            ev.set_seq(seq);
            if (out1Writer.write(ev, 0L) != 0 || out2Writer.write(ev, 0L) != 0) {
                System.err.println("FAIL: write() failed at seq=" + seq);
                System.exit(1);
            }
        }
        System.out.println("Session: wrote " + SAMPLE_COUNT + " samples on SessionOut1/SessionOut2");

        int received = 0;
        Dcps.DDS.Duration_t waitStep = new Dcps.DDS.Duration_t();
        waitStep.set_sec(WAIT_STEP_SEC);
        waitStep.set_nanosec(0);
        int waitedMs = 0;
        while (received < RECEIVE_TARGET) {
            if (waitedMs >= RECEIVE_TIMEOUT_MS) {
                System.err.println("FAIL: session did not receive from peer within " + (RECEIVE_TIMEOUT_MS / 1000) + "s (got " + received + ")");
                System.exit(1);
            }
            java.util.ArrayList<Dcps.DDS.Condition> active = new java.util.ArrayList<>();
            int wr = ws.wait(active, waitStep);
            if (wr == Dcps.DDS.RETCODE_TIMEOUT.value) {
                waitedMs += WAIT_STEP_SEC * 1000;
                continue;
            }
            if (wr != 0) {
                System.err.println("FAIL: WaitSet.wait() returned " + wr);
                System.exit(1);
            }
            SessionEventDataReader.Sample[] samples = inReader.take_n(
                SAMPLE_COUNT, Dcps.DDS.ANY_SAMPLE_STATE.value, Dcps.DDS.ANY_VIEW_STATE.value, Dcps.DDS.ANY_INSTANCE_STATE.value);
            for (SessionEventDataReader.Sample s : samples) {
                if (s.validData) received++;
            }
        }
        System.out.println("Session: received " + received + " samples from peer.");

        // The core test: tear the whole tree down in one shot instead of
        // deleting out1Dw/out2Dw/inDr/cftDr/cft one at a time.
        tornDown.set(true);

        int rc1 = dp.delete_contained_entities();
        if (rc1 != Dcps.DDS.RETCODE_OK.value) {
            System.err.println("FAIL: delete_contained_entities() returned " + rc1 + ", expected RETCODE_OK");
            System.exit(1);
        }

        int rc2 = factory.delete_participant(dp);
        if (rc2 != Dcps.DDS.RETCODE_OK.value) {
            System.err.println("FAIL: delete_participant() returned " + rc2 + " after delete_contained_entities -- cascade left something dangling");
            System.exit(1);
        }

        System.out.println("Session: torn down via delete_contained_entities.");
    }
}

File: integration-tests/cpp/delete-contained-entities/src/session.cpp
/*
 * integration-tests/cpp/delete-contained-entities -- session. Direct C++
 * port of c/delete-contained-entities/src/session.c -- see that file's
 * header comment for the full scenario rationale and
 * docs/design/integration-test-tier.md for the scenario spec.
 *
 * Required stdout markers: "Create topic:" x3, "Create writer for topic:"
 * x2, "Create reader for topic:" x2, "Session: torn down via
 * delete_contained_entities." Any failure path prints a line starting
 * "FAIL:" and exits nonzero.
 */
#include "session_event.hpp"
#include "zzdds_cpp.hpp"
#include "dcps_impl.hpp"

#include <atomic>
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <memory>
#include <unistd.h>

namespace {

constexpr int SAMPLE_COUNT = 5;
constexpr int RECEIVE_TARGET = 2;
constexpr int READER_READY_TIMEOUT_MS = 10000;
constexpr int RECEIVE_TIMEOUT_MS = 20000;
constexpr ::DDS::Duration_t WAIT_STEP{1, 0};
constexpr int POLL_PERIOD_MS = 20;

std::atomic<bool> torn_down{false};

struct WriterSyncState {
    std::atomic<bool> reader_ready{false};
};

class WriterListener : public ::zzdds::DataWriterListenerExBase {
public:
    explicit WriterListener(WriterSyncState *state) : state_(state) {}

    void on_publication_matched(std::shared_ptr<::DDS::DataWriter> /*writer*/,
                                 ::DDS::PublicationMatchedStatus /*status*/) override {
        if (torn_down.load()) {
            std::fprintf(stderr, "FAIL: listener fired after delete_contained_entities\n");
            std::exit(1);
        }
    }

    void on_reliable_reader_ready(::DDS::InstanceHandle_t /*reader_handle*/, bool is_ready) override {
        if (is_ready) state_->reader_ready.store(true);
    }

private:
    WriterSyncState *state_;
};

class ReaderListener : public ::DDS::DataReaderListenerBase {
public:
    void on_subscription_matched(std::shared_ptr<::DDS::DataReader> /*reader*/,
                                  ::DDS::SubscriptionMatchedStatus /*status*/) override {
        if (torn_down.load()) {
            std::fprintf(stderr, "FAIL: listener fired after delete_contained_entities\n");
            std::exit(1);
        }
    }
};

uint32_t parse_domain(int argc, char **argv) {
    for (int i = 1; i < argc - 1; i++) {
        if (std::strcmp(argv[i], "-d") == 0 || std::strcmp(argv[i], "--domain") == 0) {
            return static_cast<uint32_t>(std::strtoul(argv[i + 1], nullptr, 10));
        }
    }
    return 0;
}

} // namespace

int main(int argc, char **argv) {
    uint32_t domain_id = parse_domain(argc, argv);

    auto factory = zzdds::create_factory();
    if (!factory) {
        std::fprintf(stderr, "FAIL: createFactory() failed\n");
        return 1;
    }

    auto dp = factory->create_participant(domain_id, ::DDS::DomainParticipantQos::default_value(), nullptr, 0);
    if (!dp) {
        std::fprintf(stderr, "FAIL: create_participant() failed on domain %u\n", domain_id);
        return 1;
    }
    auto dp_handle = dp->native_handle();

    if (SessionEventTypeSupport::register_type(dp_handle) != 0) {
        std::fprintf(stderr, "FAIL: register SessionEventTypeSupport failed\n");
        return 1;
    }

    auto out1_topic = dp->create_topic("SessionOut1", "SessionEvent", ::DDS::TopicQos::default_value(), nullptr, 0);
    auto out2_topic = dp->create_topic("SessionOut2", "SessionEvent", ::DDS::TopicQos::default_value(), nullptr, 0);
    auto in_topic = dp->create_topic("SessionIn", "SessionEvent", ::DDS::TopicQos::default_value(), nullptr, 0);
    if (!out1_topic || !out2_topic || !in_topic) {
        std::fprintf(stderr, "FAIL: create_topic() failed\n");
        return 1;
    }
    std::printf("Create topic: SessionOut1\n");
    std::printf("Create topic: SessionOut2\n");
    std::printf("Create topic: SessionIn\n");

    auto pub = dp->create_publisher(::DDS::PublisherQos::default_value(), nullptr, 0);
    if (!pub) {
        std::fprintf(stderr, "FAIL: create_publisher() failed\n");
        return 1;
    }

    auto dw_qos = ::DDS::DataWriterQos::default_value();
    dw_qos.reliability.kind = ::DDS::ReliabilityQosPolicyKind::RELIABLE_RELIABILITY_QOS;
    dw_qos.history.kind = ::DDS::HistoryQosPolicyKind::KEEP_ALL_HISTORY_QOS;

    auto out1_dw = pub->create_datawriter(out1_topic, dw_qos, nullptr, 0);
    auto out2_dw = pub->create_datawriter(out2_topic, dw_qos, nullptr, 0);
    if (!out1_dw || !out2_dw) {
        std::fprintf(stderr, "FAIL: create_datawriter() failed\n");
        return 1;
    }
    std::printf("Create writer for topic: SessionOut1\n");
    std::printf("Create writer for topic: SessionOut2\n");

    WriterSyncState out1_state, out2_state;
    auto out1_listener = std::make_shared<WriterListener>(&out1_state);
    auto out2_listener = std::make_shared<WriterListener>(&out2_state);
    auto zout1_dw = std::static_pointer_cast<::zzdds::DataWriterImpl>(out1_dw);
    auto zout2_dw = std::static_pointer_cast<::zzdds::DataWriterImpl>(out2_dw);
    if (zout1_dw->set_listener_ex(out1_listener, DDS_PUBLICATION_MATCHED_STATUS) != ::DDS::RETCODE_OK ||
        zout2_dw->set_listener_ex(out2_listener, DDS_PUBLICATION_MATCHED_STATUS) != ::DDS::RETCODE_OK)
    {
        std::fprintf(stderr, "FAIL: set_listener_ex (writer) failed\n");
        return 1;
    }

    auto sub = dp->create_subscriber(::DDS::SubscriberQos::default_value(), nullptr, 0);
    if (!sub) {
        std::fprintf(stderr, "FAIL: create_subscriber() failed\n");
        return 1;
    }

    auto dr_qos = ::DDS::DataReaderQos::default_value();
    dr_qos.reliability.kind = ::DDS::ReliabilityQosPolicyKind::RELIABLE_RELIABILITY_QOS;
    dr_qos.history.kind = ::DDS::HistoryQosPolicyKind::KEEP_ALL_HISTORY_QOS;

    auto zin_topic = std::static_pointer_cast<::zzdds::TopicImpl>(in_topic);
    auto in_desc = zin_topic->as_topic_description();
    auto in_dr = sub->create_datareader(in_desc, dr_qos, nullptr, 0);
    if (!in_dr) {
        std::fprintf(stderr, "FAIL: create_datareader(SessionIn) failed\n");
        return 1;
    }
    std::printf("Create reader for topic: SessionIn\n");

    // Exercise CFT-collateral cleanup under the cascade -- this project has
    // a real CFT bug history (see docs/decisions.md). Trivial filter: just
    // needs to exist and be attached, not to actually narrow anything.
    auto cft = dp->create_contentfilteredtopic("SessionIn_cft", in_topic, "seq >= 0", {});
    if (!cft) {
        std::fprintf(stderr, "FAIL: create_contentfilteredtopic() failed\n");
        return 1;
    }
    auto cft_dr = sub->create_datareader(cft, dr_qos, nullptr, 0);
    if (!cft_dr) {
        std::fprintf(stderr, "FAIL: create_datareader(SessionIn_cft) failed\n");
        return 1;
    }
    std::printf("Create reader for topic: SessionIn_cft\n");

    auto reader_listener = std::make_shared<ReaderListener>();
    if (in_dr->set_listener(reader_listener, DDS_SUBSCRIPTION_MATCHED_STATUS) != ::DDS::RETCODE_OK ||
        cft_dr->set_listener(reader_listener, DDS_SUBSCRIPTION_MATCHED_STATUS) != ::DDS::RETCODE_OK)
    {
        std::fprintf(stderr, "FAIL: set_listener (reader) failed\n");
        return 1;
    }

    auto ws = zzdds::create_waitset();
    if (!ws) {
        std::fprintf(stderr, "FAIL: create_waitset() failed\n");
        return 1;
    }
    auto in_rc = in_dr->create_readcondition(::DDS::ANY_SAMPLE_STATE, ::DDS::ANY_VIEW_STATE, ::DDS::ANY_INSTANCE_STATE);
    if (!in_rc) {
        std::fprintf(stderr, "FAIL: create_readcondition() failed\n");
        return 1;
    }
    if (ws->attach_condition(in_rc) != ::DDS::RETCODE_OK) {
        std::fprintf(stderr, "FAIL: attach_condition() failed\n");
        return 1;
    }

    SessionEventDataWriter out1_writer(out1_dw->native_handle());
    SessionEventDataWriter out2_writer(out2_dw->native_handle());
    SessionEventDataReader in_reader(in_dr->native_handle());

    // Gate writes on the peer's reader actually being registered, not just
    // matched -- see docs/design/integration-test-tier.md's raw-loan/
    // coherent-sets precedent and this project's on_reliable_reader_ready
    // work: matched-count alone (SEDP discovery) does not imply the remote
    // RELIABLE reader proxy has registered this writer yet.
    for (int waited_ms = 0; !(out1_state.reader_ready.load() && out2_state.reader_ready.load()); waited_ms += POLL_PERIOD_MS) {
        if (waited_ms >= READER_READY_TIMEOUT_MS) {
            std::fprintf(stderr, "FAIL: no reliable reader became ready within %ds\n", READER_READY_TIMEOUT_MS / 1000);
            return 1;
        }
        usleep(POLL_PERIOD_MS * 1000);
    }

    for (int seq = 0; seq < SAMPLE_COUNT; seq++) {
        ::SessionEvent ev;
        ev.seq = seq;
        if (out1_writer.write(ev) != 0 || out2_writer.write(ev) != 0) {
            std::fprintf(stderr, "FAIL: write() failed at seq=%d\n", seq);
            return 1;
        }
    }
    std::printf("Session: wrote %d samples on SessionOut1/SessionOut2\n", SAMPLE_COUNT);

    int received = 0;
    int waited_ms = 0;
    while (received < RECEIVE_TARGET) {
        if (waited_ms >= RECEIVE_TIMEOUT_MS) {
            std::fprintf(stderr, "FAIL: session did not receive from peer within %ds (got %d)\n", RECEIVE_TIMEOUT_MS / 1000, received);
            return 1;
        }
        ::DDS::ConditionSeq active;
        auto wr = ws->wait(active, WAIT_STEP);
        if (wr == ::DDS::RETCODE_TIMEOUT) {
            waited_ms += 1000;
            continue;
        }
        if (wr != ::DDS::RETCODE_OK) {
            std::fprintf(stderr, "FAIL: WaitSet.wait() returned %d\n", static_cast<int>(wr));
            return 1;
        }
        std::vector<::SessionEvent> values(SAMPLE_COUNT);
        std::vector<DDS_SampleInfo> infos(SAMPLE_COUNT);
        int n = in_reader.take_n(values.data(), infos.data(), SAMPLE_COUNT,
                                  ::DDS::ANY_SAMPLE_STATE, ::DDS::ANY_VIEW_STATE, ::DDS::ANY_INSTANCE_STATE);
        for (int i = 0; i < n; i++) {
            if (infos[i].valid_data) received++;
        }
    }
    std::printf("Session: received %d samples from peer.\n", received);

    // The core test: tear the whole tree down in one shot instead of
    // deleting out1_dw/out2_dw/in_dr/cft_dr/cft one at a time.
    torn_down.store(true);

    auto rc1 = dp->delete_contained_entities();
    if (rc1 != ::DDS::RETCODE_OK) {
        std::fprintf(stderr, "FAIL: delete_contained_entities() returned %d, expected RETCODE_OK\n", static_cast<int>(rc1));
        return 1;
    }

    auto rc2 = factory->delete_participant(dp);
    if (rc2 != ::DDS::RETCODE_OK) {
        std::fprintf(stderr, "FAIL: delete_participant() returned %d after delete_contained_entities -- "
                              "cascade left something dangling\n", static_cast<int>(rc2));
        return 1;
    }

    std::printf("Session: torn down via delete_contained_entities.\n");
    return 0;
}
